DEVDOCS-5010 [new]: GQL Storefront API, complexity header (#2019)
bigcommerce/dev-docs: docs/api-docs/storefront/graphql/graphql-api-overview.mdx

## Complexity limits

> Complexity scores estimate the load that a single GraphQL request may generate. BigCommerce calculates the complexity score using an algorithm before executing the call, based on the provided query.

Reviewer: ```suggestion
Complexity scores estimate the load that a single GraphQL request can generate. BigCommerce calculates the complexity score using an algorithm before executing the call, based on the provided query.
```

The API sets a complexity limit of 10,000 for each request to prevent excessive loads. Because all work is generated per request, the limit is fixed per request, regardless of the number of requests sent.

When you send a valid GraphQL request, the API returns the query complexity as an integer using the HTTP response header `X-Bc-Graphql-Complexity` .

> If a query's complexity score exceeds the complexity limit, you will receive an error response similar to the following: 

Reviewer: ```suggestion
If a query's complexity score exceeds the complexity limit, an error response similar to the following appears: 
```

}
```

> A query's complexity may also exceed the limit due to high query depth.

Reviewer 2: I'd frame it a little differently; technically speaking, it's a separate limit.
```suggestion
In addition to the complexity limit, there is a query depth limit. The limit is X
```
@bc-dima-pasieka could you check what the depth limit is?

Reviewer 3: Oops I missed it. The current depth limit is 16

Author: Thanks @bc-dima-pasieka . One more question, would it be accurate to say that my example has a query depth of 12? Ir not, what does depth mean?

```
query {                                // depth 1
  site {                                  // depth 2
    categoryTree {                 // depth 3
      ...CategoryFields            // depth 4
      children {                       // depth 5
        ...CategoryFields          // depth 6
        children {                     // depth 7
          ...CategoryFields        // depth 8
          children {                   // depth 9
            ...CategoryFields      // depth 10
            children {                 // depth 11
              ...CategoryFields    // depth 12
            }
          }
        }
      }
    }
  }
}
```

Reviewer 3: Not quite.
```
query {                                // depth 0
  site {                                  // depth 1
    categoryTree {                 // depth 2
      ...CategoryFields            // depth 3
      children {                       // depth 3
        ...CategoryFields          // depth 4
        children {                     // depth 4
          ...CategoryFields        // depth 5
          children {                   // depth 5
            ...CategoryFields      // depth 6
            children {                 // depth 6
              ...CategoryFields    // depth 7
            }
          }
        }
      }
    }
  }
}
```
Even though we didn't reach the depth limit in this particular case, the query would fail because of the complexity limit. Another example is:
```
query { // depth 0
  site { // depth 1
    products { // depth 2
      edges { // depth 3
        node { // depth 4
          entityId // depth 5
          variants { // depth 5
            edges { // depth 6
              node { // depth 7
                entityId // depth 8
                options { // depth 8
                  edges { // depth 9
                    node { // depth 10
                      entityId // depth 11
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

```

Reviewer 3: basically depth is added by every parent type. 
In this example both nodes are on the same level
```
...CategoryFields 
children {
```

Author: Thanks @bc-dima-pasieka ! I updated the examples

}
```

> Filter arguments and character limits do not contribute to complexity.

Reviewer: ```suggestion
Filter arguments and character limits don't contribute to complexity.
```

```graphql filename="Example query" showLineNumbers copy
query {
  site {
    content {
      blog {
        posts (filters: {tags:["Most Popular"]}) { // Filter argument
          edges {
            node {
              plainTextSummary (characterLimit: 100) // Character limit
            }
          }
        }
      }
    }
  }
}
```

To reduce complexity, reduce the number of objects requested. For example, do the following:
* Limit collections to a smaller page size (for example, `first:10` instead of `first:50`).
* Reduce the number of items in nested collections, exponentially affecting complexity.
* Request fewer fields on objects, and don't request fields you don't need.

> Mutations have higher complexity than queries; you may only send one mutation in a single request. [Pagination info](#pagination) also contributes to complexity.

Reviewer: ```suggestion
Mutations have a higher complexity than queries; you can only send one mutation in a single request. [Pagination info](#pagination) also contributes to complexity.
```

### Query depth

> In addition to the complexity limit, there is a query depth limit. The limit is 16. Queries must not exceed the query depth limit and the complexity limit.

Reviewer: ```suggestion
There is a query depth limit, In addition to the complexity limit. The limit is 16. Queries must not exceed the query depth limit and the complexity limit.
```
